// @flatfile/orm/dist/index.mjs
// src/RecordMutation.ts
var RecordMutation = class {
  constructor(recordOrId) {
    this.$edits = {};
    this.$info = [];
    if (recordOrId instanceof FlatfileRecord) {
      this.recordId = recordOrId.recordId;
    } else {
      this.recordId = recordOrId;
    }
  }
  set(field, value) {
    this.$edits[field] = value;
    return this;
  }
  addInfo(fields, message) {
    return this.addMessage(fields, message, "info" /* INFO */);
  }
  addComment(fields, message) {
    return this.addInfo(fields, message);
  }
  addError(fields, message) {
    return this.addMessage(fields, message, "error" /* ERROR */);
  }
  addWarning(fields, message) {
    return this.addMessage(fields, message, "warn" /* WARN */);
  }
  addMessage(fields, message, level) {
    fields = Array.isArray(fields) ? fields : [fields];
    fields.forEach((key) => {
      this.$info.push({
        key,
        message,
        level
      });
    });
    return this;
  }
  toGraphQLEdits() {
    return {
      _id: this.recordId,
      data: this.$edits,
      messages: this.$info.map((i) => ({ error: i.level, key: i.key, message: i.message }))
    };
  }
};

// src/FlatfileRecord.ts
var FlatfileRecord = class {
  constructor($raw) {
    this.$raw = $raw;
    this.data = $raw.data;
    this.recordId = $raw.id;
    this.$info = this.$raw.info.map((i) => ({
      level: i.level,
      field: i.key,
      message: i.message
    }));
  }
  get valid() {
    return this.$raw.valid;
  }
  get status() {
    return this.$raw.status;
  }
  get allMessages() {
    return this.$info;
  }
  get errors() {
    return this.$info.filter((m) => m.level === ELevel.ERROR);
  }
  get warnings() {
    return this.$info.filter((m) => m.level === ELevel.WARN);
  }
  get info() {
    return this.$info.filter((m) => m.level === ELevel.INFO);
  }
  getMutation() {
    return new RecordMutation(this);
  }
};
var ELevel = /* @__PURE__ */ ((ELevel2) => {
  ELevel2["ERROR"] = "error";
  ELevel2["WARN"] = "warn";
  ELevel2["INFO"] = "info";
  return ELevel2;
})(ELevel || {});
var ERecordStatus = /* @__PURE__ */ ((ERecordStatus2) => {
  ERecordStatus2["REVIEW"] = "review";
  ERecordStatus2["DISMISSED"] = "dismissed";
  ERecordStatus2["ACCEPTED"] = "accepted";
  return ERecordStatus2;
})(ERecordStatus || {});

// src/RecordError.ts
var RecordError = class extends RecordMutation {
  constructor(recordOrId, errors) {
    super(recordOrId);
    errors.forEach((e) => {
      this.addError(e.field, e.message);
    });
  }
};

// src/data/basic.ts
var DATA_RECORD_BASIC = {
  id: 9,
  valid: true,
  status: "review" /* REVIEW */,
  data: {
    full_name: "John Doe",
    email: "john@doe.com",
    zip_code: "2121"
  },
  info: [
    {
      level: "info" /* INFO */,
      key: "full_name",
      message: "That's a cool name, John"
    },
    {
      level: "warn" /* WARN */,
      key: "email",
      message: "That email looks kinda fake"
    },
    {
      level: "error" /* ERROR */,
      key: "zip_code",
      message: "Not a real zip code"
    }
  ]
};
export {
  DATA_RECORD_BASIC,
  ELevel,
  ERecordStatus,
  FlatfileRecord,
  RecordError,
  RecordMutation
};
